Feature/programming exercises/trigger build button + new submission type instructor (rated after due date passes)

thilo-behnke opened the pull request:
<!-- Thanks for contributing to ArTEMiS! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
- [x] I tested the changes and all related features on the test server https://artemistest.ase.in.tum.de.
- [x] I documented my source code using the JavaDoc / JSDoc style.
- [x] I added integration test cases for the server (Spring) related to the features
- [x] I added integration test cases for the client (Jest) related to the features
- [x] I added screenshots/screencast of my UI changes
- [x] I translated all the newly inserted strings

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->

Currently it is not possible to retrigger a build without committing onto the repository.
However there are cases where this would be better, e.g. when the solution participation build should be triggered again.

### Description
<!-- Describe your changes in detail -->

**Implementation:**
- Trigger button as self contained client components
- There is one wrapper component for a normal "MANUAL" build run (normal, manual submission) and one for the "INSTRUCTOR" build run (special submission that is also considered rated after the due date passes) - I wanted to separate the endpoints on the server
- The instructor trigger button is always shown, the student trigger only if the last submission has not received a result within 2 minutes.
- When clicked, starts a build for the given participation's repo on Bamboo
- The button is disabled until the result comes back from Bamboo (submission websocket service)
- :warning: The button is disabled if there is no result yet for the participation (this would usually  mean that the student has not made a commit on the repository yet, which would introduce a new scenario)

Out of scope for this PR (will be tackled in a follow up):
- Add "trigger all" button in participation list
- Handle accumulated errors on pages instead of showing alerts (e.g. exercise-row could show multiple alerts on the same page)

**Places where the new button is shown atm:**
- [x] Programming Exercise Detail
- [x] Course exercise row
- [x] Participation list (/#/exercise/8/participation) (Note: It might be useful to also have a 'trigger all' button here - however as far as I see this comes with further complications and didn't want to make this PR even bigger)
- [x] course-exercise-detail (/#/overview/13/exercises/596)
- [x] Code-Editor

**Other changes that are not directly related to the topic of the PR:**
- Fixed programming exercise server tests that would seemingly non-deterministicly fail (tearDown db after and not before tests, fixed a couple of bugs when creating the programming exercises in the tests)
- Streamlined the `programming-exercise-participations` rest urls

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->

1. Log in to ArTEMiS
2. Go to courses > exercises / course administration > programming-exercise view
3. Click on trigger button and watch how the build is executed :)

### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->

**Course exercise row showing the tooltip:**

![course-row-error-trigger](https://user-images.githubusercontent.com/28230611/63211411-e3550b00-c0f6-11e9-93f9-20f1f0833ddb.png)

**Programming exercise detail with running build:**

![exercise-detail-running](https://user-images.githubusercontent.com/28230611/62855470-9cc87080-bcf2-11e9-829b-b428e9539794.png)

**Participation list:**

![trigger-participations](https://user-images.githubusercontent.com/28230611/63088416-2b3c2c80-bf55-11e9-9ff3-cca180c20a9b.png)

krusche commented:
I guess we could leave out the button title to make it shorter

thilo-behnke commented:
There were 3 issues with the tests for programming exercises:
- In some cases, the solution/template participation was not saved into the correct repository
- The order of assigning the solution/template participation to an exercise was incorrect (first the participation needs to be saved, then attached to the exercise)
- Most important: The database reset needs to be done in `@AfterEach`. I noticed that sometimes the data from a previous test could not be deleted from the repositories.

thilo-behnke commented on src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingSubmissionResource.java:
```diff
@@ -95,6 +104,68 @@ public ProgrammingSubmissionResource(ProgrammingSubmissionService programmingSub
+     */
+    @PostMapping(Constants.PROGRAMMING_SUBMISSION_RESOURCE_PATH + "{participationId}/trigger-build")
+    @PreAuthorize("hasAnyRole('USER', 'TA', 'INSTRUCTOR', 'ADMIN')")
+    public ResponseEntity<Void> triggerBuild(@PathVariable Long participationId) {
```
There is some duplication here, however I didn't want to mix up the normal build and the instructor build for safety's sake.
Otherwise we would have to identify within the endpoint who the user is and thereby set the corresponding submissionType (Manual vs. Instructor). This would also make it impossible for the instructor to execute a manual build for their own participation (which might sometimes be useful).

krusche commented on src/main/java/de/tum/in/www1/artemis/domain/Result.java:
```diff
@@ -294,6 +295,21 @@ public void setRatedIfNotExceeded(ZonedDateTime exerciseDueDate, ZonedDateTime s
+        if (submission.getType() == SubmissionType.INSTRUCTOR || submission.getType() == SubmissionType.TEST) {
+            this.rated = true;
+        }
+        else
```
please always use curly brackets

thilo-behnke replied:
I changed it! I also checked if there is a formatting rule for this in Spotless, but apparently not: https://help.eclipse.org/mars/index.jsp?topic=%2Forg.eclipse.jdt.doc.isv%2Freference%2Fapi%2Forg%2Feclipse%2Fjdt%2Fcore%2Fformatter%2FDefaultCodeFormatterConstants.html

krusche commented on src/main/webapp/i18n/de/programmingExercise.json:
```diff
@@ -74,7 +74,9 @@
                 "weightCouldNotBeUpdated": "Die Gewichtung für den Testfall {{id}} konnte nicht übernommen werden: {{error}}."
-            }
+            },
+            "resubmit": "Erneute Abgabe deiner Lösung. Diese Aktion führt einen neuen Build aus, der deine Submission testet.",
```
```suggestion
            "resubmit": "Erneute Abgabe deiner Lösung. Diese Aktion führt einen neuen Build aus, der deine Abgabe testet.",
```

krusche commented:
I started to review but I need some more time. Please fix the conflicts in between.
What I was wondering: Should we really always show this trigger button? Or should we only show it in certain scenarios?

To be discussed:
* I think we can leave it out for students if the build and all tests passed (i.e. 100%)
* I am not sure if students need to see it for failed tests all the time? Maybe only in certain cases, e.g. when there is a submission but no result (after 2 minutes) or when there are certain feedback messages (e.g. time out exceptions)

What do you think? @sleiss @ungaralex (also it would be great to get another review from you)

thilo-behnke commented:
> I started to review but I need some more time. Please fix the conflicts in between.
> What I was wondering: Should we really always show this trigger button? Or should we only show it in certain scenarios?
> 
> To be discussed:
> 
> * I think we can leave it out for students if the build and all tests passed (i.e. 100%)
> * I am not sure if students need to see it for failed tests all the time? Maybe only in certain cases, e.g. when there is a submission but no result (after 2 minutes) or when there are certain feedback messages (e.g. time out exceptions)
> 
> What do you think? @sleiss @ungaralex (also it would be great to get another review from you)

You are right, I think the student does not need to see the button if the last submission received a result. **So the only case where the button would be shown is if the last submission did not receive a result.**
I will adjust the feature accordingly and ask for another review when I'm done.

krusche commented on src/main/java/de/tum/in/www1/artemis/service/ProgrammingSubmissionService.java:
```diff
@@ -1,8 +1,11 @@
 import java.time.ZonedDateTime;
 import java.util.Optional;
 
+import org.eclipse.jgit.api.errors.GitAPIException;
```
I was wondering whether we should really expose these classes outside of the GitService. I think from a system design point of view (low coupling and high cohesion), it would be better, if only the GitService has dependencies to org.eclipse.jgit.* if possible.

thilo-behnke replied:
I agree; the issue is with the way the GitService was implemented it exposes the GitApiException in its methods instead of providing more information on what the actual issue is or means.

I will change it for the `getLastCommitHash` method.

thilo-behnke replied:
Done.

krusche commented on src/main/java/de/tum/in/www1/artemis/service/ProgrammingSubmissionService.java:
```diff
@@ -126,4 +133,29 @@ public ProgrammingSubmission getLatestPendingSubmission(Long participationId) th
+        URL repoUrl = participation.getRepositoryUrlAsUrl();
+        ObjectId lastCommitHash;
+        try {
+            lastCommitHash = gitService.getLastCommitHash(repoUrl);
```
This basically means that we would have 2 programming submissions with the same commit hash, right?
When the result from Bamboo is retrieved, we search for the programming submission with a specific commit hash. What happens then in case we find two submission with the same commit hash?

Are you sure that there can be no problems in the ResultService / BambooService in such a case?

thilo-behnke replied:
I think there is no other way to handle it, if we don't want to make an empty commit.

The Bamboo.onBuildCompletedNew fetches the pending submissions ordered by submission date desc and then tries to find the first one with a matching commit hash:

```
            List<ProgrammingSubmission> submissions = programmingSubmissionRepository.findByParticipationIdAndResultIsNullOrderBySubmissionDateDesc(participation.getId());
            Optional<ProgrammingSubmission> latestMatchingPendingSubmission = submissions.stream().filter(submission -> {
                String matchingCommitHashInBuildMap = getCommitHash(buildMap, submission.getType());
                return matchingCommitHashInBuildMap.equals(submission.getCommitHash());
            }).findFirst();
```

So the "worst case scenario" would be:
- Student executes a submission, the build is created on Bamboo
- The build takes longer than 2 minutes, this enables the student to trigger the submission again
- A new submission with the same submission is created on the server, there are now 2 submissions for the same commitHash and 2 running builds
- The first build returns a result to Artemis, this result is now attached to the second submission (that was just created)
- The second build finishes and returns a result to Artemis, this result is attached to the first submission

thilo-behnke replied:
(I also renamed the method to `createSubmissionWithLastCommitHashForParticipation` to make the intention more clear.

krusche replied:
I think this is fine, as long as we do not get severe problems by this approach. The worst case scenario sounds ok to me, but I would suggest that you add this as comment into the source code. I would write it as detailed as possible so that future developers are aware of it

thilo-behnke replied:
You are right, I changed to JavaDoc:

```
    /**
     * Create a submission with given submission type for the last commit hash of the given participation.
     * WARNING: The commitHash is used to map incoming results to submissions. Using this method could cause the result to have multiple fitting submissions.
     *
     * @param participation to create submission for.
     * @param submissionType of the submission to create.
     * @return created submission.
     * @throws IllegalStateException if the last commit hash can't be retrieved.
     */
```

krusche replied:
Can you put the whole comment of this thread into the JavaDoc as discussion? even the worst case scenario?

thilo-behnke replied:
I added a link to this conversion and the worst case scenario to the docs.

thilo-behnke commented:
@krusche @ungaralex @sleiss 

**Changelog since the last review:**
- Open discussion point from Stephan: https://github.com/ls1intum/Artemis/pull/712#discussion_r314944129
- Adjusted the student build trigger button (course-exercise-row):
  - Only show if the last submission does not have a result and is older than 2 minutes
  - I updated the tooltip accordingly and made the button **red** as it is an important indicator for the student that something went wrong (see screenshot above in description)
  - The instructor trigger button (participation list, programming exercise view) is always visible
  - TBD: Contrary to Stephan's suggestion, I would still show the trigger button if the user has a submission with a 100% result.  (https://github.com/ls1intum/Artemis/pull/712#issuecomment-521777729)

krusche commented:
I have on additional request, but this might be better for another pull request:
Imports to `org.eclipse.jgit.*` should only occur in the `GitService.java`, `Repository.java` and `RepositoryExecutor.java` (which should all be moved into a git package below service). 

This means the import should **NOT** occur in any other file (except in test cases). I just searched for the import and it occurs in around 10 files. I think this will improve the system design. And while it is not planned to change the git library, this would be easier in the future in case it might be necessary.

thilo-behnke commented:
> I have on additional request, but this might be better for another pull request:
> Imports to `org.eclipse.jgit.*` should only occur in the `GitService.java`, `Repository.java` and `RepositoryExecutor.java` (which should all be moved into a git package below service).
> 
> This means the import should **NOT** occur in any other file (except in test cases). I just searched for the import and it occurs in around 10 files. I think this will improve the system design. And while it is not planned to change the git library, this would be easier in the future in case it might be necessary.

I agree, this would be a bigger change to the whole `GitService` as it doesn't handle most of the exceptions atm and just hands out the GitApiExceptions. It is not a trivial change as for every thrown GitApiException we will have to decide how we want to handle it (does it mean an EntityNotFound? or IllegalArgumentException?).

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingSubmissionResource.java:
```diff
@@ -95,6 +104,68 @@ public ProgrammingSubmissionResource(ProgrammingSubmissionService programmingSub
+     *
+     * @param participationId of the participation.
+     * @return ok if the participation could be found and has permissions, otherwise forbidden (403) or notFound (404). Will also return notFound if the user's git repository is not available.
+     */
```
```suggestion
     * The REST path would be: "/programming-submissions/{participationId}/trigger-build"
     */
```

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingSubmissionResource.java:
```diff
@@ -95,6 +104,68 @@ public ProgrammingSubmissionResource(ProgrammingSubmissionService programmingSub
+     *
+     * @param participationId of the participation.
+     * @return ok if the participation could be found and has permissions, otherwise forbidden (403) or notFound (404). Will also return notFound if the user's git repository is not available.
+     */
```
```suggestion
     * The REST path would be: "/programming-submissions/{participationId}/trigger-instructor-build"
     */
```

krusche commented on src/main/webapp/app/entities/programming-exercise/actions/programming-exercise-trigger-build-button.component.html:
```diff
@@ -0,0 +1,9 @@
+<button
+    *ngIf="alwaysShowTriggerButton || (participationHasResult && participationIsActive && participationHasLatestSubmissionWithoutResult)"
```
I am wondering if `participationHasResult` is really necessary here. What happens if `participationHasLatestSubmissionWithoutResult` evaluates to true, but there was no result yet?

thilo-behnke replied:
You are right, now that we only show the button when there is a failed submission, it doesn't make sense anymore. I removed this logic from the component.

krusche commented:
When I open the participation list of a course for the first time, I see the following exception error
<img width="906" alt="image" src="https://user-images.githubusercontent.com/744067/63266137-8a6caa80-c28f-11e9-9952-74101f453095.png">

Why does it occur?

thilo-behnke commented:
> When I open the participation list of a course for the first time, I see the following exception error
> <img alt="image" width="906" src="https://user-images.githubusercontent.com/744067/63266137-8a6caa80-c28f-11e9-9952-74101f453095.png">
> 
> Why does it occur?

I agree, this message doesn't make much sense here. The implementation of the build process visualization in https://github.com/ls1intum/Artemis/pull/707 makes the submission service directly emit an error to the jhiAlert system. This doesn't always make sense though.
I removed the alert from all other places than the code-editor for now as the alert doesn't seem to be a good solution in most cases. I think we could tackle this in the follow-up PR (e.g. show an accumulated alert in the participation list or mark a row in the exercise-row component red if it has a failed submission to make the error more obvious).

thilo-behnke commented:
@krusche Thanks for your detailed review!

Changelog since last review:
- The trigger button is now also red in the participation list IF the last submission failed
- The exercise-detail page now also uses the updating-result component and shows the trigger button next to it
- The "submission failed" alerts are disabled from all pages except the code-editor
- Fixed missing resources in manage test cases due to wrong merge in translation file
